python/feathub/feature_tables/tests/test_mysql_source_sink.py

```diff
@@ -0,0 +1,222 @@
+#  Copyright 2022 The Feathub Authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from abc import ABC
+from datetime import datetime
+from typing import List
+
+import numpy as np
+import pandas as pd
+import sqlalchemy
+from pandas._testing import assert_frame_equal
+
+from feathub.common import types
+from feathub.feature_tables.sinks.mysql_sink import MySQLSink
+from feathub.feature_tables.sources.mysql_source import MySQLSource
+from feathub.feature_views.on_demand_feature_view import OnDemandFeatureView
+from feathub.table.schema import Schema
+from feathub.tests.feathub_it_test_base import FeathubITTestBase
+
+
+class MySQLSourceSinkITTest(ABC, FeathubITTestBase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.setup_mysql_container()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.teardown_mysql_container()
+
+    def test_mysql_sink_upsert(self):
+        self.client = self.get_client(
+            extra_config={"processor": {"flink": {"native.parallelism.default": "1"}}}
+        )
+
+        table_name = self.generate_random_name("test_mysql_sink_upsert")
+        engine = sqlalchemy.create_engine(self.mysql_container.get_connection_url())
+        with engine.begin() as conn:
+            conn.execute(
+                sqlalchemy.text(
+                    f"CREATE TABLE {table_name} ("
+                    f"id BIGINT PRIMARY KEY, val BIGINT, ts TEXT)"
+                )
+            )
+
+        input_data = pd.DataFrame(
+            [
+                [1, 1, datetime(2022, 1, 1, 0, 0, 0).strftime("%Y-%m-%d %H:%M:%S")],
+                [2, 2, datetime(2022, 1, 1, 0, 0, 1).strftime("%Y-%m-%d %H:%M:%S")],
+                [3, 3, datetime(2022, 1, 1, 0, 0, 2).strftime("%Y-%m-%d %H:%M:%S")],
+                [1, 4, datetime(2022, 1, 1, 0, 0, 3).strftime("%Y-%m-%d %H:%M:%S")],
+            ],
+            columns=["id", "val", "ts"],
+        )
+        schema = (
+            Schema.new_builder()
+            .column("id", types.Int64)
+            .column("val", types.Int64)
+            .column("ts", types.String)
+            .build()
+        )
+        source = self.create_file_source(
+            input_data,
+            keys=["id"],
+            schema=schema,
+            timestamp_field="ts",
+            timestamp_format="%Y-%m-%d %H:%M:%S",
+        )
+        sink = self._get_mysql_sink(table_name)
+        self.client.materialize_features(
```

Comment:
It seems that the logic here pretty much duplicate the logic in `_initialize_mysql_table()`. Should we re-use _initialize_mysql_table()?

It might be reasonable to remove this test if the logic in this test is a subset of the logic in `test_get_online_features()`.

Reply:
I update the test of MySQLSink to reuse the `_initialize_mysql_table` method. But I think we should keep both tests, as they test different aspects and use cases.